flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java

```diff
@@ -835,8 +840,21 @@ public void updateAccumulators(final AccumulatorSnapshot accumulatorSnapshot) {
                         mainThreadExecutor);
     }
 
-    private void startCheckpointScheduler(final CheckpointCoordinator checkpointCoordinator) {
+    @Override
+    public void stopCheckpointScheduler() {
+        Preconditions.checkState(
+                getCheckpointCoordinator() != null,
+                "Checkpointing cannot be stopped since it's not enabled.");
+        getCheckpointCoordinator().stopCheckpointScheduler();
+    }
+
+    @Override
+    public void startCheckpointScheduler() {
         mainThreadExecutor.assertRunningInMainThread();
+        final CheckpointCoordinator checkpointCoordinator = getCheckpointCoordinator();
+        Preconditions.checkState(
+                checkpointCoordinator != null,
+                "Checkpointing cannot be started since it's not enabled.");
```

Comment:
The checkpoint scheduler can be null, in situation I consider valid:
scenario:
- job with no restarts configured
- savepoint creation fails

- result: the EG reaches a terminal state, and sets the checkpoint coordinator to null. This precondition fails, killing the JobManager process.

Reply:
I agree: That's a valid scenario. I'm just wondering now whether we should simply make the call optional in that case. We could just add another condition to the if statement. I don't like that because it makes the interface less specific. On the other hand one could argue that starting the checkpoint scheduling was always optional due to the condition that checks whether periodic checkpointing is enabled.

Handling it We cannot just wait for the termination to restart the scheduling due to the happy path. Another alternative would be to extend the interface by a method to check the state of the checkpointing (like `CheckpointCoordinator.shutdown` and only re-enable it if it's not shutdown. @rmetzger any thoughts on this.

Reply:
I went for the first option. Unfortunately, I had to remove the `CheckpointScheduling` interface from the `CheckpointCoordinator` due to that since the `CheckpointCoordinator` is throwing an exception in case of a shutdown.

The approach with having a method for checking whether Checkpoint Scheduling is enabled wasn't better either because it would have meant: Adding this check would have exposed the shutdown behavior of the `CheckpointCoordinator` which, as a consequence, would have required to add the `shutdown` method to the interface as well. This method doesn't work for the Scheduler, though. Hence, I decided to go for the approach that checks the availability of the `CheckpointCoordinator` internally.